src/Runtime/OMTensor.inc

```diff
@@ -0,0 +1,434 @@
+      _offset = 0;
+      _dataType = ONNX_TYPE_UNDEFINED;
+      _rank = rank;
+      _owning = false;
+    } else {
+      throw std::runtime_error(
+          "OMTensor(" + std::to_string(rank) + ") malloc error");
+    }
+  };
+
+  OMTensor() = default;
+
+  /**
+   * Destructor
+   *
+   * Destroy the OMTensor struct.
+   */
+  ~OMTensor() {
+    if (_owning)
+      free(_allocatedPtr);
+    free(_shape);
+    free(_stride);
+  };
+#endif
+  // Fields are named according to:
+  // https://mlir.llvm.org/docs/Dialects/SPIR-V/#lowering-memrefs-to-spvarray-and-spvrtarray
+
+  void *_allocatedPtr; // data buffer
+  void *_alignedPtr;   // aligned data buffer that the omt indexes.
+
+  int64_t _offset;  // offset of 1st element
+  int64_t *_shape;  // sizes array
+  int64_t *_stride; // strides array
+  int64_t _rank;    // rank
+
+  OM_DATA_TYPE _dataType; // ONNX data type
+
+  int _owning; // indicates whether the Omt owns the memory space
+               // referenced by _allocatedPtr. Omt struct will release the
+               // memory space referred to by _allocatedPtr upon destruction if
+               // and only if it owns it.
+};
+
+// Create a OMTensor.
+OMTensor *omTensorCreate(
+    void *data_ptr, int64_t *shape, int64_t rank, OM_DATA_TYPE dtype) {
+  OMTensor *tensor = (OMTensor *)malloc(sizeof(OMTensor));
+  if ((tensor->_shape = (int64_t *)malloc(rank * sizeof(int64_t))) &&
+      (tensor->_stride = (int64_t *)malloc(rank * sizeof(int64_t)))) {
+    tensor->_allocatedPtr = data_ptr;
+    tensor->_alignedPtr = data_ptr;
+    tensor->_rank = rank;
+    tensor->_dataType = dtype;
+    tensor->_owning = false;
+  }
+  // Using signed indices helps detect when index falls below 0.
+  for (int64_t i = rank - 1; i >= 0; i--) {
+    tensor->_shape[i] = shape[i];
+    if (i == rank - 1)
+      tensor->_stride[i] = 1;
+    else
+      tensor->_stride[i] = tensor->_stride[i + 1] * tensor->_shape[i + 1];
+  }
+  return tensor;
+}
+
+// Create a OMTensor.
+OMTensor *omTensorCreateWithOwnership(void *data_ptr, int64_t *shape,
+    int64_t rank, OM_DATA_TYPE dtype, int owning) {
+  OMTensor *tensor = omTensorCreate(data_ptr, shape, rank, dtype);
+  tensor->_owning = owning;
+  return tensor;
+}
+
+// Create a OMTensor.
+OMTensor *omTensorCreateEmptyDeprecated(int rank) {
+  OMTensor *omt = (OMTensor *)malloc(sizeof(struct OMTensor));
+  if ((omt->_shape = (int64_t *)malloc(rank * sizeof(int64_t))) &&
+      (omt->_stride = (int64_t *)malloc(rank * sizeof(int64_t)))) {
+    omt->_allocatedPtr = NULL;
+    omt->_alignedPtr = NULL;
+    omt->_offset = 0;
+    omt->_dataType = ONNX_TYPE_UNDEFINED;
+    omt->_rank = rank;
+    omt->_owning = false;
+  }
+  return omt;
+}
+
+OMTensor *omTensorCreateEmpty(
+    int64_t *shape, int64_t rank, OM_DATA_TYPE dtype) {
+  OMTensor *tensor =
+      omTensorCreateWithOwnership(NULL, shape, rank, dtype, true);
+  void *dataPtr = malloc(omTensorGetNumElems(tensor) * getDataTypeSize(dtype));
```

Comment:
Need to check for `omTensorCreateWithOwnership` and `malloc` failure. And if `malloc` fails, tensor must be destroyed.